### PR TITLE
New Github Issues Spice

### LINKED ISSUE #3170: Github: Extend IA to search issues
### Description

Based on [user feedback](https://github.com/duckduckgo/zeroclickinfo-spice/issues/3105), we're not doing a good enough job at surfacing relevant Github issues. I propose we extend the existing Github Instant Answer to search for issues or creating a new one. The first approach, although not perfect, would be to list issues from the Github API - https://api.github.com/search/issues?q=rails%20invalid%20authenticity and list relevant issues to the user. 

Currently, there are two separate IAs. One for the [Github status](https://duck.co/ia/view/github_status) and another for [Github Projects](https://duck.co/ia/view/github). 

## Get Started
- [x] 1) Claim this issue by commenting below
- [x] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/CONTRIBUTING.md)
- [x] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

Instant Answer Page: https://duck.co/ia/view/github

### PR DESCRIPTION
**Pull request by ManrajGrover**
## Description of new Instant Answer, or changes
Github Issues Spice allows searching Github for Issues related to the query

## Related Issues and Discussions
Fixes #3170 

## People to notify
@pjhampton @moollaza 

---

Instant Answer Page: https://duck.co/ia/view/github_issues

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Spice/Github.pm [hunk 1]
```diff
@@ -19,7 +19,7 @@ spice proxy_cache_valid => '200 1d';
 
 handle query_lc => sub {
     s/^github\s*|\s+github$//;
-    if ($_ eq "" || m/\bjobs\b|\bstatus\b/) {
+    if ($_ eq "" || m/\bjobs\b|\bstatus\b|\bissue\b/) {
```
Comment (moollaza) [inline lib/DDG/Spice/Github.pm]: @ManrajGrover please put these changes in another PR. We need to keep PRs to 1 IA each 👍

**Comment by ManrajGrover** [comment]
@moollaza This can be reviewed

File: share/spice/github_issues/github_issues.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                tile: 'basic4',
+            }
+        };
+        
+        DDG.require("moment.js", function() {
+            Spice.add({
+                id: 'github_issues',
+                name: 'Github Issues',
+                data: results,
+                meta: metaObj,
```
Comment (moollaza) [inline share/spice/github_issues/github_issues.js]: might as well skip the variable and define this in-place

File: share/spice/github_issues/github_issues.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        var templateObj = {};
+        var metaObj = { 
+            sourceName: 'GitHub',
+            searchTerm: decodedQuery,
+            itemType: itemType,
+            rerender: [ 'description', 'image' ],
+            sourceUrl: "https://github.com/search?type=Issues&q=" + encodeURIComponent(query)
+        };
+        
+        templateObj = {
```
Comment (moollaza) [inline share/spice/github_issues/github_issues.js]: might as well skip the variable and define this in-place, below

File: lib/DDG/Spice/GithubIssues.pm [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+spice is_cached => 1;
+spice proxy_cache_valid => '200 1d';
+
+spice wrap_jsonp_callback => 0;
+
+spice to => 'https://api.github.com/search/issues?q=$1&callback={{callback}}';
+
+my @languages = share("languages.txt")->slurp;
+chomp(@languages);
+my $langs = join("|", map(quotemeta, @languages));
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: You need to sort the languages by decreasing length to ensure longer strings are captured first. e.g. "coldfusion" vs "coldfusion cfc" -- that might be the only case though...

**Comment by moollaza** [comment]
Deploying to Beta 👍

**Comment by ManrajGrover** [comment]
@moollaza This can be reviewed again 😄

File: lib/DDG/Spice/GithubIssues.pm [hunk 5]
```diff
@@ -14,31 +14,28 @@ spice wrap_jsonp_callback => 0;
...
 # Handle statement
-handle query_lc => sub {
-    s/github issue//;
-    
-    if ($_ eq "") {
-        return;
-    }
-    
+handle remainder_lc => sub {
+    return unless (($_) or $_ ne "");
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: No need for all the parens. Also if $_ is defined, it's guaranteed to have a length, so you don't need to check that its not `""` 👍 

`return unless $_;` should be fine.

File: lib/DDG/Spice/GithubIssues.pm [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package DDG::Spice::GithubIssues;
+# ABSTRACT: Search Github for Issues related to the query
+
+use DDG::Spice;
+use strict;
+use warnings;
+use Text::Trim;
+
+spice is_cached => 1;
+spice proxy_cache_valid => '200 1d';
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: maybe we should drop this to `1h`? caching for a day seems a little long

File: lib/DDG/Spice/GithubIssues.pm [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+use Text::Trim;
+
+spice is_cached => 1;
+spice proxy_cache_valid => '200 1h';
+
+spice wrap_jsonp_callback => 0;
+
+spice to => 'https://api.github.com/search/issues?q=$1&callback={{callback}}';
+
+my @languages = share("languages.txt")->slurp;
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: You can slurp and chomp at the same time 👍 

`->slurp( chomp => 1 )`

File: lib/DDG/Spice/GithubIssues.pm [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+my $langs = join("|", map(quotemeta, @languages));
+
+triggers startend => 'github issue';
+
+# Handle statement
+handle remainder_lc => sub {
+    return unless $_;
+
+    my $query = $_;
+    $query =~ s/\b($langs)\b//;
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: we need to quote the names so that `+` and other characters aren't interpreted in the regex:

`\b(\Q$langs\E)\b`

this ensures a query like 'github issue c++ linux` matches on `c++`, instead of `c` :+1:

File: lib/DDG/Spice/GithubIssues.pm [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+spice to => 'https://api.github.com/search/issues?q=$1&callback={{callback}}';
+
+my @languages = share("languages.txt")->slurp;
+@languages = sort { length($b) <=> length($a) } @languages;
+
+chomp(@languages);
+my $langs = join("|", map(quotemeta, @languages));
+
+triggers startend => 'github issue';
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: I think we need to make this an `end` trigger only. Looking at some queries, none of the ones that start with `github issue` seem to be a search for specific issues

File: lib/DDG/Spice/GithubIssues.pm (hunk 8 above)
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: we need to quote the names so that `+` and other characters aren't interpreted in the regex:

`\b(\Q$langs\E)\b`

this ensures a query like 'github issue c++ linux' matches on `c++`, instead of `c` :+1:

File: lib/DDG/Spice/GithubIssues.pm [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+handle remainder_lc => sub {
+    return unless $_;
+
+    my $query = $_;
+    $query =~ s/\b($langs)\b//;
+    trim($query);
+
+    my $lang;
+
+    if (/\b($langs)\b/) {
```
Comment (moollaza) [inline lib/DDG/Spice/GithubIssues.pm]: Why not do this first?

```perl
handle remainder_lc => {
  return unless $_;
  if (m/.../) {
    $lang = $1;
    s/.../;
    return qq($_ language:"$lang");
  }

  return $_;
}
```

File: lib/DDG/Spice/GithubIssues.pm (hunk 8 above)
Comment (ManrajGrover) [inline lib/DDG/Spice/GithubIssues.pm]: Looks like `\Q` and `\E` treats pipes as word literal. http://rextester.com/FIOPX95634

**Comment by moollaza** [comment]
@ManrajGrover Thanks! This LGTM :+1:

**Comment by daxtheduck** [comment]
@ManrajGrover this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!
